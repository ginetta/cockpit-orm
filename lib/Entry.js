'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _lodash = require('lodash.get');

var _lodash2 = _interopRequireDefault(_lodash);

var _lodash3 = require('lodash.set');

var _lodash4 = _interopRequireDefault(_lodash3);

var _isEmpty = require('is-empty');

var _isEmpty2 = _interopRequireDefault(_isEmpty);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _has = function _has(x) {
  return !(0, _isEmpty2.default)(x);
};

var Entry = function () {
  function Entry(_ref) {
    var _this = this;

    var cockpit = _ref.cockpit,
        schema = _ref.schema,
        fields = _ref.fields;

    _classCallCheck(this, Entry);

    this.Map = new Map();
    this.defaultSchema = ['_by', '_created', '_id', '_mby', '_modified', '_order'];

    this.cockpit = cockpit;
    this.fields = fields;
    this.schema = schema;

    if (!this.fields) return;
    if (!this.schema) return;

    var slugFields = this.schema.fields.filter(function (_ref2) {
      var options = _ref2.options;
      return options && options.slug;
    }).map(function (_ref3) {
      var name = _ref3.name;
      return { name: name + '_slug' };
    });

    var possibleFields = [].concat(_toConsumableArray(schema.fields), _toConsumableArray(slugFields), _toConsumableArray(this.defaultSchema.map(function (name) {
      return { name: name };
    })));

    possibleFields.forEach(function (_ref4) {
      var name = _ref4.name;

      _this.Map.set(name, _this.fields[name]);
    });
  }

  _createClass(Entry, [{
    key: 'get',
    value: function get(field, path) {
      if (path) {
        return (0, _lodash2.default)(this.Map.get(field), path);
      }

      return this.Map.get(field);
    }
  }, {
    key: 'has',
    value: function has(field, path) {
      if (path) {
        return _has(this.Map.get(field), path);
      }

      return _has(this.Map.get(field));
    }
  }, {
    key: 'set',
    value: function set(field, value, path) {
      if (path) {
        var sett = (0, _lodash4.default)(this.get(field), path, value);

        return this.Map.set(sett);
      }

      this.Map.set(field, value);

      return this;
    }
  }, {
    key: 'setSlug',
    value: function setSlug(slug) {
      var slugField = this.constructor.slugField;


      this.set(slugField, slug);

      return this;
    }
  }, {
    key: 'forEach',
    value: function forEach(fn) {
      return this.Map.forEach(fn);
    }
  }, {
    key: 'getImageField',
    value: function getImageField(field, options) {
      return this.cockpit.image(this.get(field, 'path'), options);
    }
  }, {
    key: 'sync',
    value: function sync() {
      var _this2 = this;

      var slugField = this.constructor.slugField;

      var filter = _defineProperty({}, slugField, this.get(slugField));

      var collection = this.cockpit.collection(this.constructor.collectionName, { filter: filter });

      return collection.promise.then(_utils.entriesHead).then(function (entry) {
        _this2.Map = new _this2.constructor(entry).Map;

        return _this2;
      });
    }
  }, {
    key: 'watch',
    value: function watch(callback) {
      var _this3 = this;

      var slugField = this.constructor.slugField;

      var filter = {};

      if (slugField) filter[slugField] = this.get(slugField);
      if (this.has('_id')) filter._id = this.get('_id');

      var collection = this.cockpit.collection(this.constructor.collectionName, { filter: filter });

      collection.watch(function (entry) {
        _this3.Map = new _this3.constructor((0, _utils.entriesHead)(entry)).Map;

        callback(_this3);
      });

      collection.on('preview', function (entry) {
        _this3.Map = new _this3.constructor(entry.data).Map;

        callback(_this3);
      });

      return this;
    }
  }, {
    key: 'toObject',
    value: function toObject() {
      return [].concat(_toConsumableArray(this.Map)).reduce(function (prev, curr) {
        return Object.assign({}, prev, _defineProperty({}, curr[0], curr[1]));
      }, {});
    }
  }]);

  return Entry;
}();

Entry.slugField = 'name_slug';
exports.default = Entry;