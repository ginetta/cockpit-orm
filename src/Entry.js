import get from 'lodash.get';
import set from 'lodash.set';
import isEmpty from 'is-empty';
import { entriesHead } from './utils';

const has = x => !isEmpty(x);

export default class Entry {
  static slugField = 'name_slug';

  Map = new Map();

  defaultSchema = ['_by', '_created', '_id', '_mby', '_modified', '_order'];

  constructor({ cockpit, schema, fields }) {
    this.cockpit = cockpit;
    this.fields = fields;
    this.schema = schema;

    if (!this.fields) return;
    if (!this.schema) return;

    const slugFields = this.schema.fields
      .filter(({ options }) => options && options.slug)
      .map(({ name }) => ({ name: `${name}_slug` }));

    const possibleFields = [
      ...schema.fields,
      ...slugFields,
      ...this.defaultSchema.map(name => ({ name })),
    ];

    possibleFields.forEach(({ name }) => {
      this.Map.set(name, this.fields[name]);
    });
  }

  get(field, path) {
    if (path) {
      return get(this.Map.get(field), path);
    }

    return this.Map.get(field);
  }

  has(field, path) {
    if (path) {
      return has(this.Map.get(field), path);
    }

    return has(this.Map.get(field));
  }

  set(field, value, path) {
    if (path) {
      const sett = set(this.get(field), path, value);

      return this.Map.set(sett);
    }

    this.Map.set(field, value);

    return this;
  }

  setSlug(slug) {
    const { slugField } = this.constructor;

    this.set(slugField, slug);

    return this;
  }

  forEach(fn) {
    return this.Map.forEach(fn);
  }

  getImageField(field, options) {
    return this.cockpit.image(this.get(field, 'path'), options);
  }

  sync() {
    const { slugField } = this.constructor;
    const filter = { [slugField]: this.get(slugField) };

    const collection = this.cockpit.collection(
      this.constructor.collectionName,
      { filter },
    );

    return collection.promise.then(entriesHead).then(entry => {
      this.Map = new this.constructor(entry).Map;

      return this;
    });
  }

  getCollection() {
    const filter = {};
    const { slugField } = this.constructor;

    if (slugField) filter[slugField] = this.get(slugField);
    if (this.has('_id')) filter._id = this.get('_id');

    return this.cockpit.collection(this.constructor.collectionName, { filter });
  }

  watch(callback) {
    const collection = this.getCollection();

    collection.watch(entry => {
      this.Map = new this.constructor(entriesHead(entry)).Map;

      callback(this);
    });

    collection.on('preview', entry => {
      this.Map = new this.constructor(entry.data).Map;

      callback(this);
    });

    return this;
  }

  save() {
    return this.cockpit.collectionSave(
      this.constructor.collectionName,
      this.toObject(),
    );
  }

  toObject() {
    return [...this.Map].reduce(
      (prev, curr) => ({
        ...prev,
        [curr[0]]: curr[1],
      }),
      {},
    );
  }
}
