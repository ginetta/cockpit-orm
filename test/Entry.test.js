import Entry from '../src/Entry';
import cockpit from './cockpit';

test('Should create without errors ', () => {
  class TestEntry extends Entry {
    static slugField = 'title_slug';
    static schema = { fields: [{ name: 'title' }] };
  }

  const entry = new TestEntry({ title: 'Foo bar' });

  expect(entry.toObject()).toMatchSnapshot();
});

test('Should get slug fields ', () => {
  class TestEntry extends Entry {
    static slugField = 'title_slug';
    static schema = { fields: [{ name: 'title', options: { slug: true } }] };
  }

  const entry = new TestEntry({ title: 'Foo bar', title_slug: 'foo-bar' });
  const slugValue = entry.get('title_slug');

  expect(slugValue).toBe('foo-bar');
});

test('Should get slug field by setting slug', () => {
  class TestEntry extends Entry {
    static slugField = 'title_slug';
    static schema = { fields: [{ name: 'title', options: { slug: true } }] };
  }

  const entry = new TestEntry({ title: 'Foo bar' });

  entry.setSlug('foo-bar');

  const slugValue = entry.get('title_slug');

  expect(slugValue).toBe('foo-bar');
  expect(entry.toObject()).toMatchSnapshot();
});

test('Should sync with slug field', () => {
  expect.assertions(3);

  class TestEntry extends Entry {
    static slugField = 'title_slug';
    static schema = {
      name: 'test',
      fields: [{ name: 'title', options: { slug: true } }],
    };
    static cockpit = cockpit;
  }

  const entry = new TestEntry({ title: 'Foo bar' });

  entry.setSlug('foo-bar');

  return entry.sync().then(value => {
    expect(value.toObject().title).toBe('Foo bar');
    expect(value.toObject()._id).not.toBe(undefined);
    expect(value.toObject()._mby).not.toBe(undefined);
  });
});

test('Should update entry', async () => {
  class TestEntry extends Entry {
    static slugField = 'title_slug';
    static schema = {
      name: 'test',
      fields: [{ name: 'title', options: { slug: true } }, { name: 'body' }],
    };
    static cockpit = cockpit;
  }

  const entry = new TestEntry().setSlug('foo-bar');

  await entry.sync();

  expect(entry.toObject().body).toBe('test body');

  entry.set('body', 'test foo');

  await entry.save();

  expect(entry.toObject().body).toBe('test foo');

  entry.set('body', 'test body');

  await entry.save();

  expect(entry.toObject().body).toBe('test body');
});

test('Should add new Entry', () => {
  expect.assertions(4);

  class TestEntry extends Entry {
    static cockpit = cockpit;
    static slugField = 'title_slug';
    static schema = {
      name: 'test',
      fields: [{ name: 'title', options: { slug: true } }, { name: 'body' }],
    };
  }

  const entry = new TestEntry({ title: 'Hello world' });

  expect(entry.get('title')).toBe('Hello world');
  expect(entry.has('body')).toBe(false);

  return entry.save().then(value => {
    expect(value.toObject().title).toBe('Hello world');
    expect(value.toObject()._id).toBeTruthy();
  });
});
