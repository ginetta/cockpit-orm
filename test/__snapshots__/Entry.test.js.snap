// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create without errors  1`] = `
Object {
  "_by": undefined,
  "_created": undefined,
  "_id": undefined,
  "_mby": undefined,
  "_modified": undefined,
  "_order": undefined,
  "title": "Foo bar",
}
`;

exports[`Should get slug field by setting slug 1`] = `
Object {
  "_by": undefined,
  "_created": undefined,
  "_id": undefined,
  "_mby": undefined,
  "_modified": undefined,
  "_order": undefined,
  "title": "Foo bar",
  "title_slug": "foo-bar",
}
`;
